crates/accelerate/src/quantum_circuit/circuit_data.rs

```diff
@@ -0,0 +1,365 @@
+            CompareOp::Ne => CircuitData::equals(slf, other).map(|r| (!r).into_py(py)),
+            _ => Ok(py.NotImplemented()),
+        }
+    }
+}
+
+enum IndexFor {
+    Lookup,
+    Insertion,
+}
+
+impl CircuitData {
+    fn convert_py_slice(&self, py: Python<'_>, slice: &PySlice) -> PyResult<Vec<isize>> {
+        let dict: HashMap<&str, PyObject> =
+            HashMap::from([("s", slice.into()), ("length", self.data.len().into_py(py))]);
+        py.eval(
+            "list(range(*s.indices(length)))",
+            None,
+            Some(dict.into_py_dict(py)),
+        )?
+        .extract()
+    }
+
+    fn convert_py_index(&self, index: isize, kind: IndexFor) -> PyResult<usize> {
+        let index = if index < 0 {
+            index + self.data.len() as isize
+        } else {
+            index
+        };
+
+        let out_of_bounds = match kind {
+            IndexFor::Lookup => index < 0 || index >= self.data.len() as isize,
+            IndexFor::Insertion => index < 0 || index > self.data.len() as isize,
+        };
+
+        if out_of_bounds {
+            return Err(PyIndexError::new_err(format!(
+                "Index {index} is out of bounds.",
+            )));
+        }
+        Ok(index as usize)
+    }
+
+    fn equals(slf: &PyAny, other: &PyAny) -> PyResult<bool> {
+        let slf_len = slf.len()?;
+        let other_len = other.len();
+        if other_len.is_ok() {
+            if slf_len != other_len.unwrap() {
+                return Ok(false);
+            }
+        }
+        let mut ours_itr = slf.iter()?;
+        let mut theirs_itr = match other.iter() {
+            Ok(i) => i,
+            Err(_) => {
+                return Ok(false);
+            }
+        };
+        loop {
+            match (ours_itr.next(), theirs_itr.next()) {
+                (Some(ours), Some(theirs)) => {
+                    if !ours?.eq(theirs?)? {
+                        return Ok(false);
+                    }
+                }
+                (None, None) => {
+                    return Ok(true);
+                }
+                _ => {
+                    return Ok(false);
+                }
+            }
+        }
+    }
+
+    fn drop_from_cache(
+        &self,
+        py: Python<'_>,
+        entry: (PyObject, IndexType, IndexType),
+    ) -> PyResult<()> {
+        let cell: &PyCell<InternContext> = self.intern_context.as_ref(py);
+        let mut py_ref: PyRefMut<'_, InternContext> = cell.try_borrow_mut()?;
+        let context = &mut *py_ref;
+        let (_, qargs_idx, carg_idx) = entry;
+        context.drop_use(qargs_idx);
+        context.drop_use(carg_idx);
+        Ok(())
+    }
+
+    fn get_or_cache(
+        &mut self,
+        py: Python<'_>,
+        elem: ElementType,
+    ) -> PyResult<(PyObject, IndexType, IndexType)> {
+        let cell: &PyCell<InternContext> = self.intern_context.as_ref(py);
+        let mut py_ref: PyRefMut<'_, InternContext> = cell.try_borrow_mut()?;
```

Comment:
Fairly minor, but `try_borrow_mut()?` is just the same as `borrow_mut()` right?  We're always panicking if we can't acquire it, but since access to it should be dependent on holding the GIL, we're ok - we can't allowing threading within functions that mutate Python-space objects anyway.

Reply:
For this, I just copied the example from the PyO3 docs:
https://pyo3.rs/v0.19.2/types#pycellsometype

I guess the difference is whether or not we panic in Rust or pass back an error to Python space. You're right that the borrow should always succeed since we've got the GIL, so it's probably reasonable to take down the process if that is violated. I don't have a strong preference.

Reply:
Oh that's true - for some reason I treated `?` in my head as the same as `Result::unwrap()`, which isn't true at all.  Ignore this comment.